### PR TITLE
Migrate WorksheetTemplates to DX

### PR DESCRIPTION
**Pull request by Bugerman58**
## Description of the issue/feature this PR addresses

This PR migrates the AT WorksheetTemplate and WorksheetTemplates to Dexterity types for Python 3 compatibility.

## Current behavior before PR

WorksheetTemplate and WorksheetTemplates are content types based on the Archetypes framework

## Desired behavior after PR is merged

WorksheetTemplate and WorksheetTemplates are content types based on the Dexterity framework

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

**Comment by Bugerman58** [comment]
Hi @ramonski and @xispa 

Sorry for a long time of prepare this PR. There were a lot of problems and questions that needed to be solved for the WST to operate under the new conditions.
There were difficulties with the `layout` field (now it's a `template_layout`) for edit and display mode.

Pls check it.

**Review by ramonski — COMMENTED** [review]
I just had a first glance at the code and must say "Wow, what a great piece of work", thanks Anton!
Can you merge the current 2.x branch please, then I will test it with some existing DBs this week.
Thanks again for your work, you are a great support guys!

**Comment by Bugerman58** [comment]
> I just had a first glance at the code and must say "Wow, what a great piece of work", thanks Anton! Can you merge the current 2.x branch please, then I will test it with some existing DBs this week. Thanks again for your work, you are a great support guys!

Thank you for your high appreciation of my work

**Comment by ramonski** [comment]
Hi Anton,
here comes a first feedback of my integration tests:

1. The last upgrade step in `senaite.core` was this one (https://github.com/senaite/senaite.core/blob/2.x/src/senaite/core/upgrade/v02_06_000.zcml#L6-L12):

``` xml
  <genericsetup:upgradeStep
      title="SENAITE.CORE 2.6.0: Remove deactivated services from profiles"
      description="Removes deactivated analysis services from Profiles"
      source="2651"
      destination="2652"
      handler=".v02_06_000.remove_inactive_services_from_profiles"
      profile="senaite.core:default"/>
```

Please update make sure you have correctly merged `2.x` and that your upgrade step is `2653`.

---

2. The current implementation adds as many rows as were specified in "Number of Positions":

Hardness — SENAITE LIMS 2024-10-15 1 PM-56-12

The migrated DX version lacks this functionality (no button available) and rows are added always on the position 1:

Worksheet Templates — SENAITE LIMS 2024-10-15 2 PM-02-48

---

3. Links to previous created positions do not work:

Bildschirmfoto 2024-10-15 um 2 05 11 PM

--- 

4. Migrating the demo data Worksheet templates where I added the layout, I get the following traceback:

```
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 176, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 385, in publish_module
  Module ZPublisher.WSGIPublisher, line 288, in publish
  Module ZPublisher.mapply, line 85, in mapply
  Module ZPublisher.WSGIPublisher, line 63, in call_object
  Module plone.autoform.view, line 40, in __call__
  Module plone.autoform.view, line 62, in _update
  Module z3c.form.group, line 52, in update
  Module z3c.form.group, line 48, in updateWidgets
  Module z3c.form.field, line 277, in update
  Module z3c.form.browser.multi, line 63, in update
  Module z3c.form.browser.widget, line 171, in update
  Module z3c.form.widget, line 496, in update
  Module Products.CMFPlone.patches.z3c_form, line 47, in _wrapped
  Module z3c.form.widget, line 132, in update
  Module z3c.form.widget, line 491, in value
  Module collective.z3cform.datagridfield.datagridfield, line 171, in updateWidgets
  Module z3c.form.widget, line 433, in updateWidgets
  Module z3c.form.widget, line 410, in applyValue
  Module collective.z3cform.datagridfield.datagridfield, line 308, in value
  Module collective.z3cform.datagridfield.datagridfield, line 262, in datagrid_field_set
  Module collective.z3cform.datagridfield.datagridfield, line 311, in updateWidgets
  Module z3c.form.object, line 208, in updateWidgets
  Module collective.z3cform.datagridfield.datagridfield, line 439, in update
  Module plone.z3cform.fieldsets.extensible, line 65, in update
  Module plone.z3cform.patch, line 30, in GroupForm_update
  Module z3c.form.group, line 132, in update
  Module collective.z3cform.datagridfield.datagridfield, line 446, in updateWidgets
  Module z3c.form.form, line 136, in updateWidgets
  Module z3c.form.field, line 277, in update
  Module z3c.form.browser.text, line 36, in update
  Module z3c.form.browser.widget, line 171, in update
  Module Products.CMFPlone.patches.z3c_form, line 47, in _wrapped
  Module z3c.form.widget, line 132, in update
  Module z3c.form.converter, line 116, in toWidgetValue
  Module zope.i18n.format, line 484, in format
TypeError: a float is required
```

## Suggestion

I would remove the "Positions" input field and provide a similar architecture as done for the partition scheme in sample templates:

Sample Templates — SENAITE LIMS 2024-10-15 2 PM-07-08

The partition selector dropdown is also dynamically updated when new partitions are added

Bildschirmfoto 2024-10-15 um 2 09 09 PM

**Comment by ramonski** [comment]
Minor Issue:

It would be great to find out, why these Analyses Tables are rendered so small on both the Worksheet- and Sample Template settings.

**Comment by ramonski** [comment]
Another UI issue:

Clicking on the "Analysis Type" Dropdown completely removes the Worksheet layout:

Worksheet Templates — SENAITE LIMS 2024-10-15 2 PM-14-01

Worksheet Templates — SENAITE LIMS 2024-10-15 2 PM-14-52

**Comment by ramonski** [comment]
Maybe related to the UI handler adapter:

When selecting a Reference, the JS Console shows this error:

Worksheet Templates — SENAITE LIMS 2024-10-15 2 PM-17-12

**Comment by ramonski** [comment]
The "Duplicate of" selector should be only visible or active if the "Analysis Type" was set to "Duplicate":

My WST — SENAITE LIMS 2024-10-15 2 PM-23-40

**Comment by Bugerman58** [comment]
Hi, Ramon! Thank for you feedback.
> Hi Anton, here comes a first feedback of my integration tests:
> 
> 2. The current implementation adds as many rows as were specified in "Number of Positions":
>
> 3. Links to previous created positions do not work:
> 
_fast answer from me_

very strange interface on your screenshots

maybe you should clear the cache(Shift+Cmd+R)

I'll attach my screenshots a bit later, but this is definitely not how it should be

**Comment by Bugerman58** [comment]
>The last upgrade step in senaite.core was this one (https://github.com/senaite/senaite.core/blob/2.x/src/senaite/core/upgrade/v02_06_000.zcml#L6-L12):
  <genericsetup:upgradeStep
      title="SENAITE.CORE 2.6.0: Remove deactivated services from profiles"
      description="Removes deactivated analysis services from Profiles"
      source="2651"
      destination="2652"
      handler=".v02_06_000.remove_inactive_services_from_profiles"
      profile="senaite.core:default"/>
Please update make sure you have correctly merged 2.x and that your upgrade step is 2653.

I'm sure that my version 2653 )

Screenshot 2024-10-15 at 19 50 00

**Comment by Bugerman58** [comment]
@ramonski 

this is the correct ui

View mode:

Screenshot 2024-10-15 at 20 10 34

Edit mode(before change):

Screenshot 2024-10-15 at 20 10 58


Edit mode(after change num positions):

Screenshot 2024-10-15 at 20 11 21

**Comment by Bugerman58** [comment]
> Suggestion
I would remove the "Positions" input field and provide a similar architecture as done for the partition scheme in sample templates:

and as for the SampleTemplates, the `partition id` field does not work correctly when adding and deleting rows:

Screenshot 2024-10-15 at 20 33 35

**Comment by ramonski** [comment]
Hi Anton,
Please apologize, I forgot that additional remotes are not fetched when doing a `git fetch` without the `--all` flag 🙈 
Ok, I'm now on [ac605ef](https://github.com/senaite/senaite.core/pull/2599/commits/ac605ef8383ead39c5f6758a2bf1e1f62bb5a87f) here and installed a fresh SENAITE site.
When importing the demo data, I get the following traceback:
```
2024-10-15 16:42:38,813 ERROR   [senaite.core:67][waitress-0] Traceback (most recent call last):
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/browser/form/adapters/data_import.py", line 62, in handle_data_import
    LoadSetupData(self.context, self.request)()
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/load_setup_data.py", line 119, in __call__
    adapter(self, workbook, self.dataset_project, self.dataset_name)
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/setupdata/__init__.py", line 144, in __call__
    self.Import()
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/setupdata/__init__.py", line 2005, in Import
    folder = self.context.bika_setup.bika_worksheettemplates
AttributeError: 'RequestContainer' object has no attribute 'bika_worksheettemplates'
```

**Comment by Bugerman58** [comment]
> Hi Anton, Please apologize, I forgot that additional remotes are not fetched when doing a `git fetch` without the `--all` flag 🙈 Ok, I'm now on [ac605ef](https://github.com/senaite/senaite.core/pull/2599/commits/ac605ef8383ead39c5f6758a2bf1e1f62bb5a87f) here and installed a fresh SENAITE site. When importing the demo data, I get the following traceback:
> 
> ```
> 2024-10-15 16:42:38,813 ERROR   [senaite.core:67][waitress-0] Traceback (most recent call last):
>   File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/browser/form/adapters/data_import.py", line 62, in handle_data_import
>     LoadSetupData(self.context, self.request)()
>   File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/load_setup_data.py", line 119, in __call__
>     adapter(self, workbook, self.dataset_project, self.dataset_name)
>   File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/setupdata/__init__.py", line 144, in __call__
>     self.Import()
>   File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/setupdata/__init__.py", line 2005, in Import
>     folder = self.context.bika_setup.bika_worksheettemplates
> AttributeError: 'RequestContainer' object has no attribute 'bika_worksheettemplates'
> ```

Excellent. You managed to achieve the correct UI behavior?

`exportimport` and `demodata` is not so important now and can be adjusted

**Comment by ramonski** [comment]
> > Suggestion
> > I would remove the "Positions" input field and provide a similar architecture as done for the partition scheme in sample templates:
> 
> and as for the SampleTemplates, the `partition id` field does not work correctly when adding and deleting rows:
> 
> Screenshot 2024-10-15 at 20 33 35

Yes, you are right. I was able to reproduce the bug and this should be definitely fixed.

**Comment by Bugerman58** [comment]
> When importing the demo data, I get the following traceback:

I'll try to fix it this week

**Comment by ramonski** [comment]
> > When importing the demo data, I get the following traceback:
> 
> I'll try to fix it this week

Great, thanks Anton! This makes testing a bit easier.

The UI looks great now and works like a charm!

My test setup has 2 instances, one with the current `2.x` branch and one with your branch:

**Old UI**

Hardness — SENAITE LIMS 2024-10-16 7 AM-53-28

**New UI**

Hardness — SENAITE LIMS 2024-10-16 7 AM-49-48

During my tests, I noted a side-effect **on both versions** when I try to use these Templates:

**Current 2.x Branch**

WS-001 — SENAITE LIMS 2024-10-16 7 AM-55-26

**Your feature Branch**

WS-001 — SENAITE LIMS 2024-10-16 7 AM-56-48

While it is good that it actually behaved the same, I asked myself why the Blank and Reference Sample were not assigned, why Position 3 is missing and Position 4 is a Reassignable slot.

In both versions there are only Reference Definitions present, which are those listed in the WST dropdowns:

Reference Definitions — SENAITE LIMS 2024-10-16 8 AM-07-49

But there are no Reference Samples below Suppliers present:

Reference Samples cc — SENAITE LIMS 2024-10-16 8 AM-08-59

@xispa: This is something we probably should address when touching Worksheets the next time.

Anyhow, just a reminder at this place for now.

**Comment by ramonski** [comment]
> Minor Issue:
> 
> It would be great to find out, why these Analyses Tables are rendered so small on both the Worksheet- and Sample Template settings.

This PR fixes the rendering issue for listing widgets: https://github.com/senaite/senaite.core/pull/2625

**Comment by ramonski** [comment]
Btw, just because I've seen it now again.
With https://github.com/senaite/senaite.core/pull/2588 you should be able to style the "Set" button like in the old version. But it is perfectly fine how it looks at the moment as well.

File: src/senaite/core/browser/form/adapters/worksheettemplate.py [hunk 1]
```diff
@@ -18,37 +18,213 @@
...
+
+        if dup_pos:
+            msg = translate(_(
+                u"duplicate_reference_this_position",
+                default=u"Duplicate in position ${dup} references this, "
+                        u"so it must be a routine analysis.",
+                mapping={"dup": dup_pos})
+            )
+            self.add_error_field(FIELD_TYPE.format(index), msg)
+            return False
```
Comment (ramonski) [inline src/senaite/core/browser/form/adapters/worksheettemplate.py:172]: Good check!

ABX Pentra 60 — SENAITE LIMS 2024-10-16 2 PM-16-58

File: src/senaite/core/content/worksheettemplate.py [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+from zope.interface import Interface
+from zope.interface import implementer
+from zope.schema.vocabulary import SimpleVocabulary
+from z3c.form.interfaces import DISPLAY_MODE
+from six.moves.urllib.parse import parse_qs
+
+
+def default_layout_positions(count_positions=None, start_pos=0):
+    """
+    """
```
Comment (ramonski) [inline src/senaite/core/content/worksheettemplate.py]: Please add a docstring here

File: src/senaite/core/content/worksheettemplate.py [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+    request = api.get_request()
+    if count_positions:
+        num_positions = int(count_positions)
+    elif request:
+        params = dict(parse_qs(request["QUERY_STRING"]))
+        values = params.get("num_positions")
+        num_positions = int(values[0]) if values else 0
```
Comment (ramonski) [inline src/senaite/core/content/worksheettemplate.py]: Add a comment here, that you are called in this case from the add form adapter (`senaite.core.browser.form.wstemplate`) with  `++add++WorksheetTemplate?num_positions={}`

File: src/senaite/core/content/worksheettemplate.py [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+            "dup": None,
+        })
+    return default_value
+
+
+def default_num_positions():
+    request = api.get_request()
+    if request:
+        params = dict(parse_qs(request["QUERY_STRING"]))
+        values = params.get("num_positions")
```
Comment (ramonski) [inline src/senaite/core/content/worksheettemplate.py]: duplicate code here. Please add a function e.g. `get_query_string_by_name(name)` into a suitable place.

**Comment by xispa** [comment]
> @xispa: This is something we probably should address when touching Worksheets the next time.

I was thinking that maybe it would only make sense to allow the selection of a sample template with References when there are counterpart QC samples registered and valid. However, the QC Samples hold reference values that apply to tests and we cannot foresee that the resulting worksheet will contain analyses for all tests defined in the template... Therefore, I think best would be to simply place an alert panel stating something like this:


> [!WARNING]  
> *This slot is intended to contain a quality control sample for the 'Trace Metals 9' reference definition; however, no active or valid quality control samples were identified.*


We could also place a reference widget with [Add] button inside that very same sample to easily add such a quality control sample when available.

This is for another PR.

File: src/senaite/core/content/worksheettemplate.py [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+    enable_multiple_use_of_instrument = schema.Bool(
+        title=_(
+            u"label_worksheettemplate_enable_multiple_user_of_instrument",
+            default=u"Enable Multiple Use of Instrument in Worksheets",
+        ),
+        description=_(
+            u"description_worksheettemplate_multiple_user_of_instrument",
+            default=u"If unchecked, Lab Managers won't be able to assign "
+                    u"the same Instrument more than one Analyses while "
+                    u"creating a Worksheet."
```
Comment (ramonski) [inline src/senaite/core/content/worksheettemplate.py]: I'm confused by the current text we have here.
@xispa: Can you remember what the purpose of this checkbox was?

File: src/senaite/core/content/worksheettemplate.py [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+    @security.protected(permissions.ModifyPortalContent)
+    def setTemplateLayout(self, value):
+        mutator = self.mutator("template_layout")
+        mutator(self, value)
+        num_pos = len(value)
+        if num_pos != self.getNumOfPositions():
+            self.setNumOfPositions(num_pos)
+
+    # BBB: AT schema field property
+    Layout = property(getTemplateLayout, setTemplateLayout)
```
Comment (ramonski) [inline src/senaite/core/content/worksheettemplate.py:455]: Please add the method `getLayout` with a deprecation decorator for backwards compatibility here

File: src/senaite/core/content/worksheettemplate.py [hunk 7]
```diff
@@ -0,0 +1,570 @@
...
+                    "Expected object, uid or record, got %r" % type(v))
+            records.append({
+                "uid": uid,
+            })
+
+        mutator = self.mutator("services")
+        mutator(self, records)
+
+    # BBB: AT schema field property
+    Services = property(getServices, setServices)
```
Comment (ramonski) [inline src/senaite/core/content/worksheettemplate.py:527]: Please add the method `getRawService` with a deprecation decorator for backwards compatibility here

File: src/senaite/core/content/worksheettemplate.py (hunk 5 above)
Comment (ramonski) [reply src/senaite/core/content/worksheettemplate.py]: Ah ok, found it here:
https://github.com/senaite/senaite.core/blob/2.x/src/bika/lims/browser/worksheet/templates/results.pt#L255-L258
And I guess it has absolutely no function, or at least I cannot find a JS that queries the ID `instrument_multiple_use`

Shall we simply remove this field?

File: src/senaite/core/z3cform/widgets/datagrid/datagrid.py [hunk 8]
```diff
@@ -66,6 +67,11 @@ def createObjectWidget(self, idx):
 
         return widget
 
+    def render(self):
+        if self.mode in self.templates.keys():
+            return ViewPageTemplateFile(self.templates[self.mode])(self)
+        return super(DataGridWidget, self).render()
+
```
Comment (ramonski) [inline src/senaite/core/z3cform/widgets/datagrid/datagrid.py:74]: cool!

File: src/senaite/core/z3cform/widgets/datagrid/ws_template_datagrid_display.pt [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+              <div tal:replace="structure pos/render"></div>
+            </td>
+            <td>
+              <div tal:replace="structure type/render"></div>
+            </td>
+            <td>
+              <?python
+                blank_ref.mode = "display"
+                control_ref.mode = "display"
+              ?>
```
Comment (ramonski) [inline src/senaite/core/z3cform/widgets/datagrid/ws_template_datagrid_display.pt:48]: Oh wow, I haven't seen/used this construct before!

**Review by ramonski — CHANGES_REQUESTED** [review]
Great work Anton and I see the efforts you have put into this PR!
I added some comments (also for @xispa) and I guess when we have the updated setupdata, then we can merge it into Core!

File: src/senaite/core/content/worksheettemplate.py (hunk 5 above)
Comment (xispa) [reply src/senaite/core/content/worksheettemplate.py]: Pheew, this comes from #133 ... really old code... and not sure what was the expected functionality... Yes, remove the field, please

File: src/senaite/core/z3cform/widgets/datagrid/datagrid.py (hunk 8 above)
Comment (Bugerman58) [reply src/senaite/core/z3cform/widgets/datagrid/datagrid.py:74]: Tnx! )

File: src/senaite/core/content/worksheettemplate.py (hunk 5 above)
Comment (Bugerman58) [reply src/senaite/core/content/worksheettemplate.py]: I agree. Done.

File: src/senaite/core/content/worksheettemplate.py (hunk 4 above)
Comment (Bugerman58) [reply src/senaite/core/content/worksheettemplate.py]: oh, sure.

File: src/senaite/core/browser/form/adapters/worksheettemplate.py (hunk 1 above)
Comment (Bugerman58) [reply src/senaite/core/browser/form/adapters/worksheettemplate.py:172]: Thanks! Idea and text  copy-paste from old code  🤫

File: src/senaite/core/z3cform/widgets/datagrid/ws_template_datagrid_display.pt (hunk 9 above)
Comment (Bugerman58) [reply src/senaite/core/z3cform/widgets/datagrid/ws_template_datagrid_display.pt:48]: nano-hack for template of widget 😉

File: src/senaite/core/content/worksheettemplate.py (hunk 6 above)
Comment (Bugerman58) [reply src/senaite/core/content/worksheettemplate.py:455]: Hi Ramon!
Can not declare method `getLayout` because this method invoke from parent Plone view and throw exception:
```shell
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 176, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 385, in publish_module
  Module ZPublisher.WSGIPublisher, line 264, in publish
  Module ZPublisher.BaseRequest, line 450, in traverse
  Module plone.rest.patches, line 15, in __before_publishing_traverse__
  Module Products.CMFCore.DynamicType, line 164, in __before_publishing_traverse__
  Module Products.CMFDynamicViewFTI.fti, line 239, in queryMethodID
  Module Products.CMFDynamicViewFTI.fti, line 201, in defaultView
  Module Products.CMFPlone.PloneTool, line 696, in browserDefault
  Module bika.lims, line 195, in wrapper
TypeError: getLayout() got an unexpected keyword argument 'check_exists'
```

[`Products.CMFPlone.PloneTool:695`](https://github.com/plone/Products.CMFPlone/blob/master/Products/CMFPlone/PloneTool.py#L695)

`bika.lims, line 195` is deprecation decorator

_this example from my instance_

Comment (Bugerman58) [reply src/senaite/core/content/worksheettemplate.py:455]: That is why the `layout` field was renamed to `template_layout`: https://github.com/senaite/senaite.core/pull/2599#issuecomment-2376098725

Comment (Bugerman58) [reply src/senaite/core/content/worksheettemplate.py:455]: without annotation of deprecation:
```shell
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 176, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 385, in publish_module
  Module ZPublisher.WSGIPublisher, line 264, in publish
  Module ZPublisher.BaseRequest, line 450, in traverse
  Module plone.rest.patches, line 15, in __before_publishing_traverse__
  Module Products.CMFCore.DynamicType, line 164, in __before_publishing_traverse__
  Module Products.CMFDynamicViewFTI.fti, line 239, in queryMethodID
  Module Products.CMFDynamicViewFTI.fti, line 201, in defaultView
  Module Products.CMFPlone.PloneTool, line 696, in browserDefault
TypeError: getLayout() got an unexpected keyword argument 'check_exists'
```
...the same

Comment (Bugerman58) [reply src/senaite/core/content/worksheettemplate.py:455]: @ramonski pls, check that is above

**Comment by ramonski** [comment]
Thanks for the changes Anton!
I tested the migration of an existing Database and everything seems to work without issues.
Unfortunately, the demo data import still raises an error:

```
2024-10-21 20:21:51,307 ERROR   [senaite.core:67][waitress-1] Traceback (most recent call last):
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/browser/form/adapters/data_import.py", line 62, in handle_data_import
    LoadSetupData(self.context, self.request)()
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/load_setup_data.py", line 119, in __call__
    adapter(self, workbook, self.dataset_project, self.dataset_name)
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/setupdata/__init__.py", line 144, in __call__
    self.Import()
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/setupdata/__init__.py", line 2033, in Import
    obj.setTemplateLayout(self.wst_layouts[title])
AttributeError: Worksheet_Templates instance has no attribute 'wst_layouts'
```

**Comment by Bugerman58** [comment]
> Thanks for the changes Anton! I tested the migration of an existing Database and everything seems to work without issues. Unfortunately, the demo data import still raises an error:
> 
> ```
> 2024-10-21 20:21:51,307 ERROR   [senaite.core:67][waitress-1] Traceback (most recent call last):
>   File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/browser/form/adapters/data_import.py", line 62, in handle_data_import
>     LoadSetupData(self.context, self.request)()
>   File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/load_setup_data.py", line 119, in __call__
>     adapter(self, workbook, self.dataset_project, self.dataset_name)
>   File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/setupdata/__init__.py", line 144, in __call__
>     self.Import()
>   File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/setupdata/__init__.py", line 2033, in Import
>     obj.setTemplateLayout(self.wst_layouts[title])
> AttributeError: Worksheet_Templates instance has no attribute 'wst_layouts'
> ```

oh, sorry

now checked on my instance and WST import

**Comment by ramonski** [comment]
I get now this error on my side:
```
2024-10-23 08:45:41,868 ERROR   [senaite.core:67][waitress-0] Traceback (most recent call last):
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/browser/form/adapters/data_import.py", line 62, in handle_data_import
    LoadSetupData(self.context, self.request)()
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/load_setup_data.py", line 113, in __call__
    in list(getAdapters((self.context, ), ISetupDataImporter))]
  File "/Users/rbartl/develop/buildout/eggs/zope.interface-5.5.2-py2.7-macosx-15.0-x86_64.egg/zope/interface/registry.py", line 377, in getAdapters
    adapter = factory(*objects)
  File "/Users/rbartl/develop/buildout/src/senaite.core/src/senaite/core/exportimport/setupdata/__init__.py", line 1969, in __init__
    super(Worksheet_Templates, self).__init__(context)
TypeError: super() argument 1 must be type, not classobj
```

**Comment by ramonski** [comment]
I guess it is because the base class does not inherit from `object`:
```

class WorksheetImporter:

    """Use this as a base, for normal tabular data sheet imports.
    """

    def __init__(self, context):
        self.adapter_context = context

```

**Comment by ramonski** [comment]
Ok, that seems to work. Please change the base class to this:
```python
class WorksheetImporter(object):
    """Use this as a base, for normal tabular data sheet imports.
    """
```

I'm happy when we have this Worksheet importer abandoned...

**Comment by ramonski** [comment]
The demo data worked now, but the imported worksheet template seems to be misconfigured:

image

**Comment by Bugerman58** [comment]
> I get now this error on my side:
>  super(Worksheet_Templates, self).__init__(context)
> TypeError: super() argument 1 must be type, not classobj

oh, yes. this code was added after checked import. 

> The demo data worked now, but the imported worksheet template seems to be misconfigured:

yes, I understand why this is so - the import file uses values ​​from the dictionary, not keys. I will add a check for this.

**Comment by Bugerman58** [comment]
> The demo data worked now, but the imported worksheet template seems to be misconfigured:
> 

Hi Ramon! Sorry for my inadvertence. I hope now that will working correctly. =|

File: src/senaite/core/content/worksheettemplate.py (hunk 6 above)
Comment (Bugerman58) [reply src/senaite/core/content/worksheettemplate.py:455]: ...

I think that no possibility use this method

**Comment by ramonski** [comment]
Hi Anton, Thanks for the fixtures!
I guess we're almost there and I only found the following smaller issues:

1. After demo data installation, the fields show "Constraint not satisfied". But the messages disappear after saving:

after demo-data import:

Hardness — SENAITE LIMS 2024-10-26 8 PM-29-29

after save:

Hardness — SENAITE LIMS 2024-10-26 8 PM-30-27


2. Removing the slot number and pressing the "Set" button raises this error:

Hardness — SENAITE LIMS 2024-10-26 8 PM-31-19

3. Something strange but worth to mention here as well (but probably the same thing as already mentioned above) – When I create 4 demo Samples with all the "Trace Metals" profile applied and create afterwards a WS with the Template from the demo-data, I get this result:

WS-001 — SENAITE LIMS 2024-10-26 8 PM-36-23

4. Creating a new Worksheet Template (or edit an existing) gives me this error in the JS console:




To summarize:

In my opinion, the traceback and the JS error should be fixed. The rest of the issues seem to be minor or belong to another PR. Let me know what you think (also @xispa).

**Comment by Bugerman58** [comment]
> 1. After demo data installation, the fields show "Constraint not satisfied". But the messages disappear after saving:

Solved. Added check for `blank_ref` and `control_ref` fields on value of title of definitions.

But this message(`Constraint not satisfied`) may appear for the `dup` field - this now in work.

> 2. Removing the slot number and pressing the "Set" button raises this error:

Fixed.

**Comment by xispa** [comment]
> In my opinion, the traceback and the JS error should be fixed. The rest of the issues seem to be minor or belong to another PR. Let me know what you think (also @xispa).

I Will do functional testing today. Worth to mention that I played around worksheet templates yesterday with a heavy-customized instance and I stumbled upon some issues. Will take those issues into account when playing around.

**Comment by xispa** [comment]
Except the "Constraint not satisfied" thing, the rest seems to work properly to me.

**Comment by xispa** [comment]
Just wanted to say good work @Bugerman58 , you rock!

**Comment by Bugerman58** [comment]
> Just wanted to say good work @Bugerman58 , you rock!

Thanks! 😊

**Review by ramonski — APPROVED** [review]
Thanks Anton! Besides the JS error, everything work perfectly.
I'm going to add another PR and try to fix it in the `editform.js`.
One step further towards Python 3;)
